executor/src/witgen/jit/identity_queue.rs

```diff
@@ -102,151 +130,145 @@ impl<T: FieldElement> PartialEq for QueueItem<'_, T> {
 
 impl<T: FieldElement> Eq for QueueItem<'_, T> {}
 
-/// Computes a map from each variable to the queue items it occurs in.
-fn compute_occurrences_map<'b, 'a: 'b, T: FieldElement>(
+/// Utility to compute the variables that occur in a queue item.
+/// Follows intermediate column references and employs caches.
+struct ReferencesComputer<'a, T: FieldElement> {
     fixed_data: &'a FixedData<'a, T>,
-    items: &BTreeSet<QueueItem<'a, T>>,
-) -> HashMap<Variable, Vec<QueueItem<'a, T>>> {
-    let mut intermediate_cache = HashMap::new();
-
-    // Compute references only once per identity.
-    let mut references_per_identity = HashMap::new();
-    for id in items
-        .iter()
-        .filter_map(|item| match item {
-            QueueItem::Identity(id, _) => Some(id),
-            _ => None,
-        })
-        .unique_by(|id| id.id())
-    {
-        references_per_identity.insert(
-            id.id(),
-            references_in_identity(id, fixed_data, &mut intermediate_cache),
-        );
-    }
+    intermediate_cache: HashMap<AlgebraicReferenceThin, Vec<AlgebraicReferenceThin>>,
+    /// A cache to store algebraic references in a polynomial identity, so that it
+    /// can be re-used on all rows.
+    references_per_identity: HashMap<u64, Vec<AlgebraicReferenceThin>>,
+}
 
-    items
-        .iter()
-        .flat_map(|item| {
-            let variables = match item {
-                QueueItem::Identity(id, row) => {
-                    let mut variables = references_per_identity[&id.id()]
-                        .iter()
-                        .map(|r| {
-                            let name = fixed_data.column_name(&r.poly_id).to_string();
-                            Variable::from_reference(&r.with_name(name), *row)
-                        })
-                        .collect_vec();
-                    if let Identity::BusSend(bus_send) = id {
-                        variables.extend((0..bus_send.selected_payload.expressions.len()).map(
-                            |index| {
+impl<'a, T: FieldElement> ReferencesComputer<'a, T> {
+    pub fn new(fixed_data: &'a FixedData<'a, T>) -> Self {
+        Self {
+            fixed_data,
+            intermediate_cache: HashMap::new(),
+            references_per_identity: HashMap::new(),
+        }
+    }
+    pub fn references(&mut self, item: &QueueItem<'a, T>) -> Vec<Variable> {
```

Comment:
maybe this can return an iterator, usage seems to immediately call `iter()`.

Reply:
I tried to make this return an iterator, but I got some errors wrt either references to `&mut self` or temporaries in lambda functions.